local_planner/include/local_planner/trajectory_simulator.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <eigen3/Eigen/Core>
+
+namespace avoidance {
+
+struct simulation_state {
+  float time = NAN;
+  Eigen::Vector3f position = NAN * Eigen::Vector3f::Ones();
+  Eigen::Vector3f velocity = NAN * Eigen::Vector3f::Ones();
+  Eigen::Vector3f acceleration = NAN * Eigen::Vector3f::Ones();
+};
+
+struct simulation_limits {
+  float max_z_velocity = NAN;
+  float min_z_velocity = NAN;
+  float max_xy_velocity_norm = NAN;
+  float max_acceleration_norm = NAN;
+  float max_jerk_norm = NAN;
+};
+
+std::vector<simulation_state> run_steps(const simulation_limits& config,
+                                        const simulation_state& start,
+                                        const Eigen::Vector3f& goal_direction,
+                                        float step_time, int num_steps);
+
+template <int N>
+Eigen::Matrix<float, N, 1> norm_clamp(const Eigen::Matrix<float, N, 1>& val,
+                                      float max_norm) {
+  float norm_sq = val.squaredNorm();
+  if (norm_sq > max_norm * max_norm)
+    return val * (max_norm / std::sqrt(norm_sq));
+  else
+    return val;
+}
```

Comment:
Could we move this to the cpp? Personally, I like everything that's more than just a getter or a setter to be defined in the cpp, but I guess this a matter of taste

Reply:
I want to test it, so it needs something in the header. Also, templated functions have to be in the header so that each caller can get their own instantiation depending on what they call it with.